Add new statistics to getblockstats RPC

marcinja opened the pull request:
Extends getblockstats RPC introduced in #10757.

`getblockstats` now tracks spends of specific SegWit output types: P2SH-nested P2WPKH, P2SH-nested P2WSH, native P2WPKH, and native P2WSH.  New `CTxIn` and `CScript` methods added to check that the scriptSig and  scriptPubKey of a input and the ouput it spends match the format of the above output types as specified in BIP 141. (Note: there is probably a bug in this part of the PR!) Also tracks creation of new native SegWit outputs using the same new methods.

This PR also introduces a count of transactions that signal RBF. It also adds batching, consolidation, and dust-creation metrics. 

A batching transaction is defined as having at least 3 outputs. Separately, this RPC now counts transactions that have fit into different ranges of their number of outputs.  ` //  Batch ranges =  [(1), (2), (3-4), (5-9), (10-49), (50-99), (100+)] ` These ranges are predetermined but arbitrary .

A consolidating transaction is defined as having at least 3 inputs. 

Dust creation is calculated by using the `IsDust` function on each output at different fee-rates. The outputs are then put into bins of different ranges (e.g. dust at 100 sat/vbyte or higher. These ranges are also predetermined but arbitrary .

jnewbery commented:
I'll try to review this this week if I get a chance.

@jamesob - can you also review?

jnewbery commented on src/script/script.h:
```diff
@@ -557,6 +562,7 @@ class CScript : public CScriptBase
         CScriptBase::clear();
         shrink_to_fit();
     }
+
```
nit: please remove unrelated newline addition

jnewbery commented on src/script/script.h:
```diff
@@ -532,6 +532,11 @@ class CScript : public CScriptBase
     bool IsPayToScriptHash() const;
     bool IsPayToWitnessScriptHash() const;
+    bool IsNativePayToWitnessPubKeyHash() const;
+    bool IsNativePayToWitnessScriptHash() const;
```
nit: Looks like this function is declared but not defined or used. Remove

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1693,6 +1693,24 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+            "  \"txs_spending_native_p2wsh_outputs\": xxxxx,   (numeric) The total number of transactions spending native P2WSH outputs\n"
+            "  \"new_p2wpkh_outputs\": xxxxx,   (numeric) The total number of new P2WPKH outputs\n"
+            "  \"new_p2wsh_outputs\": xxxxx,   (numeric) The total number of new P2WSH outputs\n"
+            "  \"txs_creating_p2wpkh_outputs\": xxxxx,   (numeric) The total number of new P2WPKH outputs\n"
```
Documentation doesn't match the field name (name refers to txs, doc refers to outputs).

Same for next field.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1814,24 +1915,68 @@ static UniValue getblockstats(const JSONRPCRequest& request)
             if (!g_txindex) {
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "One or more of the selected stats requires -txindex enabled");
             }
+
```
nit: this added newline should be in the previous commit.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1851,6 +1996,13 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+      dustbin_array[i+1] += dustbin_array[i];
+    }
+
+
```
nit: one newline is enough!

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1693,6 +1693,24 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+            "  \"new_p2wsh_outputs\": xxxxx,   (numeric) The total number of new P2WSH outputs\n"
+            "  \"txs_creating_p2wpkh_outputs\": xxxxx,   (numeric) The total number of new P2WPKH outputs\n"
+            "  \"txs_creating_p2wsh_outputs\": xxxxx,   (numeric) The total number of new P2WSH outputs\n"
+            "  \"txs_signalling_rbf\": xxxxx,   (numeric) The total number of transactions that signal RBF\n"
```
nit: prefer to name this `txs_signalling_opt_in_rbf`

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1770,25 +1788,108 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+    int64_t outputs_consolidated = 0;
+    int64_t txs_batching = 0;
+
+    const int CONSOLIDATION_THRESHOLD = 3;
```
nit: place these at the top of the file as constexprs

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1814,24 +1915,68 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+
                 tx_total_in += prevoutput.nValue;
                 utxo_size_inc -= GetSerializeSize(prevoutput, SER_NETWORK, PROTOCOL_VERSION) + PER_UTXO_OVERHEAD;
             }
```
You add a `TODO` in commit 2 and then remove it in commit 3. Please just remove it from commit 2 if you don't want it to be in the PR.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1881,6 +2033,34 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+    ret_all.pushKV("native_p2wsh_outputs_spent", native_p2wsh_outputs_spent);
+    ret_all.pushKV("txs_spending_nested_p2wpkh_outputs", txs_spending_nested_p2wpkh_outputs);
+    ret_all.pushKV("txs_spending_nested_p2wsh_outputs", txs_spending_nested_p2wsh_outputs);
+    ret_all.pushKV("txs_spending_native_p2wpkh_outputs", txs_spending_native_p2wpkh_outputs);
```
This is incorrect in commit 1 and then fixed in commit 3. Please just fix it in commit 1.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1770,25 +1788,108 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+
+
+                // Check if output is dust at any of the set fee rates.
+                for (int64_t i = 0; i < NUM_DUST_BINS; i++) {
```
Have you profiled how performant this is? Seems like it could be slow to run `IsDust()` many times on each output in a block.

marcinja replied:
Yup. Profiling just showed that almost all time spent in getblockstats is spent in the GetTransaction call used to get the transaction with prevoutput. But I agree that running `IsDust()` so many times here isn't the best thing to do.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -1770,25 +1788,108 @@ static UniValue getblockstats(const JSONRPCRequest& request)
 
     for (const auto& tx : block.vtx) {
-        outputs += tx->vout.size();
+        int64_t nOutputs = tx->vout.size();
```
nit: Bitcoin Core code style is camel_case for variables. Prefer `tx_outputs`.

jnewbery reviewed (commented):
Looks generally good. I can't immediately see a problem with your segwit input/output counting.

I've left a bunch of nits. A few other pointers:

- try adding a sanity check to your btcd branch that checks that swtcs <= txs_spending_native + txs_spending_nested
- even better, you could add an assert to the `getblockstats()` code here that asserts that if `tx->HasWitness()` is true, then there is at least one native segwit input or nested segwit input (you might only want that in a debug build to avoid degrading performance).
- consider adding test cases to /test/functional/rpc_getblockstats.py to cover the new functionality
- see if there are any unit tests in `/src/test` that you could modify to test your segwit counting.
- I'd split the new segwit counting code into its own commit, and then have 3 commits for the new functionality in `getblockstats`
- commit messages should have a summary line (<80 chars) and then (optionally) a more detailed log, separated by a blank line.

Commit:
Implement SegWit related stats in getblockstats

Count creation of new native P2WSH and P2WPKH outputs, and the number of
transactions creating each type of output.

Count the number of native P2WPKH, native P2WSH, nested P2WSH, and
nested P2WPKH outputs spent. Count the number of transactions that spend
each type of SegWit output.

jnewbery commented on src/primitives/transaction.cpp:
```diff
@@ -43,6 +43,40 @@ std::string CTxIn::ToString() const
 
+
+bool CTxIn::SpendsNestedPayToWitnessPubKeyHashOutput(CScript spentScriptPubKey) const {
+    if ((!spentScriptPubKey.IsPayToScriptHash()) || this->scriptWitness.IsNull()) {
```
I don't think you need the `this->` pointer derefences here. Just `scriptWitness.IsNull()` should be fine.

(same for `this->` repetitions below)

Commit:
Add stats for creation of dust outputs

For each new output, checks whether it would be considered dust at
several different fee-rates.

Commit:
Add functions to check for spent SegWit outputs

Adds member functions to CTxIn and CScript for determining if a
transaction spends any of the SegWit V0 output types.

jamesob commented on src/primitives/transaction.cpp:
```diff
@@ -43,6 +43,40 @@ std::string CTxIn::ToString() const
+}
+
+bool CTxIn::SpendsNestedPayToWitnessScriptHashOutput(CScript spentScriptPubKey) const {
+    if (!(spentScriptPubKey.IsPayToScriptHash()) || scriptWitness.IsNull()) {
```
This looks like it may be problematic; should the `!` be inside the parens? Otherwise it looks like null scriptWitnesses don't cause an immediate return.

jamesob replied:
Ah wait, nevermind - misbalanced parens when reading :)

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1693,6 +1695,24 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+            "  \"new_p2wsh_outputs\": xxxxx,   (numeric) The total number of new P2WSH outputs\n"
+            "  \"txs_creating_p2wpkh_outputs\": xxxxx,   (numeric) The total number of transactions creating new P2WPKH outputs\n"
+            "  \"txs_creating_p2wsh_outputs\": xxxxx,   (numeric) The total number of transactions new P2WSH outputs\n"
+            "  \"txs_signalling_opt_in_rbf\": xxxxx,   (numeric) The total number of transactions that signal RBF\n"
```
Not sure if noted elsewhere, but "signaling" only has one l.

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1693,6 +1695,24 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+            "  \"txs_creating_p2wpkh_outputs\": xxxxx,   (numeric) The total number of transactions creating new P2WPKH outputs\n"
+            "  \"txs_creating_p2wsh_outputs\": xxxxx,   (numeric) The total number of transactions new P2WSH outputs\n"
+            "  \"txs_signalling_opt_in_rbf\": xxxxx,   (numeric) The total number of transactions that signal RBF\n"
+            "  \"txs_consolidating\": xxxxx,   (numeric) The total number of consolidating transactions (defined as at least 3 inputs and 1 output)\n"
```
Might be worth returning a list of txids instead of a raw count, and maybe consider same for some of the other added keys. This'd bloat the size of the returned JSON by a bit, but gives much more information to the user.

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1770,25 +1790,102 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+    int64_t txs_batching = 0;
+
+    // Batch ranges =  [(1), (2), (3-4), (5-9), (10-49), (50-99), (100+)]
+    const int NUM_OUTCOUNT_BINS = 7;
```
constexpr?

jamesob commented on src/rpc/blockchain.cpp:
```diff
@@ -1814,24 +1911,67 @@ static UniValue getblockstats(const JSONRPCRequest& request)
+            }
+            if (spends_native_p2wsh_output) {
+                ++txs_spending_native_p2wsh_outputs;
+            }
```
Are these mutually exclusive conditions? Should this be absorbed into the block above?

marcinja replied:
No, the block above is for checking what kind of output each individual input spends and this block is for the entire transaction. And a transaction can spend all of the different kinds of outputs

jamesob reviewed (commented):
In general looks good, all I've got for you are nits.

I went to go find the segwit spend discrepancy [on the existing dashboard](https://dashboard.bitcoinops.org/d/xVQwtADiz/segwit-usage-1-hr-average?orgId=1) (i.e. `total_segwit_spend_percentage > sum(segwit_input_percentage)`) but I wasn't able to find it. Are we still seeing that discrepancy? Is it unique to the per-block display?

marcinja commented:
Thanks for the review. At this point, I'm pretty sure the discrepancy was either caused by a bug in the Go code that has been fixed, or a mistake in the Grafana setup. (I remade the graph at the top of that dashboard to convince myself)

Still, I'll try to add some simple tests to confirm it when I fix the remaining nits.

Commit:
Add more stats to getblockstats

Adds batching transaction counter.
Tracks the number of outputs created by each transaction.
Adds consolidating transaction counter.
Counts the number of transactions signalling opt-in RBF.

marcinja commented on src/test/script_standard_tests.cpp:
```diff
@@ -111,6 +113,8 @@ BOOST_AUTO_TEST_CASE(script_standard_Solver_success)
     BOOST_CHECK_EQUAL(solutions.size(), 1U);
     BOOST_CHECK(solutions[0] == ToByteVector(scriptHash));
+    BOOST_CHECK(s.IsPayToWitnessScriptHash());
+    BOOST_CHECK(!s.IsNativePayToWitnessPubKeyHash());
```
Added a couple of tests that verify the native SegWit functions

Commit:
Track many-to-one consolidations

getblockstats now counts the number of transactions with at least 3
inputs and exactly one output as "many-to-one" (mto) consolidations.

It also counts the total number of outputs consolidated by mto
transactions per block.

Commit:
Add value tracking stats

Tracks total value of many_to_one outputs.
Tracks total value of different kinds of SegWit outputs spent.
Tracks total value of native SegWit outputs created.

marcinja commented:
Rebased onto master with new `feerate_percentiles` and added stats for value of spent/created SegWit outputs and for total value of many-to-one consolidations.

jnewbery commented:
This PR was for review only. We're using the expand-getblockstats branch and don't need to merge into master.
